File: GeoFirestoreExample/app/src/main/java/org/imperiumlabs/geofirestoreexample/MainActivity.java
package org.imperiumlabs.geofirestoreexample;

import android.os.Bundle;
import android.util.Log;
import android.widget.*;
import java.util.*;
import com.google.firebase.firestore.*;
import org.imperiumlabs.geofirestore.*;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity implements GeoQueryDataEventListener{

    private static final String POST_COLLECTION  = "POSTS";

    private static final GeoPoint QUERY_CENTER = new GeoPoint(36.963817, -122.018284);
    private static final double QUERY_RADIUS = 5;

    private ListView listView;
    private ArrayList<String> postList = new ArrayList<>();

    private GeoFirestore geoFirestore;
    private GeoQuery geoQuery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        listView = findViewById(R.id.post_list_view);

        FirebaseFirestore db = FirebaseFirestore.getInstance();

        CollectionReference posts = db.collection(POST_COLLECTION);

        // Create a GeoFirestore instance
        this.geoFirestore = new GeoFirestore(posts);

        /*
        * Currently 5 posts exist in the database with the following descriptions and locations:
        *
        * "It is a beautiful day here at the beach!", GeoPoint(36.963719, -122.020704)
        *
        * "I just saw a huge fish while swimming!", GeoPoint(36.961182, -122.014996)
        *
        * "I want to build a sand castle as big as our home!", GeoPoint(36.963805, -122.017722)
        *
        * "I have a view of the entire city!", GeoPoint(40.748441, -73.985664)
        *
        * "This is the happiest place on Earth!", GeoPoint(33.812092, -117.918974)
        *
        */

        // Create a GeoQuery to find all posts within a radius of a certain location
        this.geoQuery = geoFirestore.queryAtLocation(QUERY_CENTER, QUERY_RADIUS);

        this.geoQuery.removeAllListeners();

        // Listen to documents regarding this GeoQuery
        this.geoQuery.addGeoQueryDataEventListener(this);
    }


    @Override
    public void onDocumentEntered(DocumentSnapshot documentSnapshot, GeoPoint location) {
        try {
            Map<String, Object> data = documentSnapshot.getData();
            String description = (String) data.get("DESCRIPTION");
            if (description != null){
                postList.add(description);
            }
        } catch (NullPointerException e) {
            Log.e("DOCUMENT_ERROR", e.getLocalizedMessage());
        } catch (ClassCastException e) {
            Log.e("DOCUMENT_ERROR", e.getLocalizedMessage());
        }
    }

    @Override
    public void onDocumentExited(DocumentSnapshot documentSnapshot) {}
    @Override
    public void onDocumentMoved(DocumentSnapshot documentSnapshot, GeoPoint location) {}
    @Override
    public void onDocumentChanged(DocumentSnapshot documentSnapshot, GeoPoint location) {}

    @Override
    public void onGeoQueryReady() {
        this.geoQuery.removeGeoQueryEventListener(this);

        ArrayAdapter adapter = new ArrayAdapter(this, android.R.layout.simple_list_item_1, postList);
        listView.setAdapter(adapter);
    }

    @Override
    public void onGeoQueryError(Exception exception) {
        Log.e("QUERY_ERROR", exception.getLocalizedMessage());
    }


}

File: GeoFirestoreExample/app/src/main/java/org/imperiumlabs/geofirestoreexample/MainActivity.kt
package org.imperiumlabs.geofirestoreexample

import android.os.Bundle
import android.util.Log
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.GeoPoint
import kotlinx.android.synthetic.main.activity_main.*
import org.imperiumlabs.geofirestore.GeoFirestore
import org.imperiumlabs.geofirestore.callbacks.GeoQueryDataEventListener
import org.imperiumlabs.geofirestore.callbacks.SingleGeoQueryDataEventListener
import org.imperiumlabs.geofirestore.extension.getLocation
import org.imperiumlabs.geofirestore.extension.setLocation

class MainActivity: AppCompatActivity() {

    companion object {
        private const val TAG = "MainActivity"
        private const val POST_COLLECTION = "POSTS"

        private val QUERY_CENTER = GeoPoint(37.7853889, -122.4056973)
        private const val QUERY_RADIUS = 5.0
    }

    private val db = FirebaseFirestore.getInstance()
    private val posts = db.collection(POST_COLLECTION)
    private val geoFirestore = GeoFirestore(posts)

    private val postList = arrayListOf<String>()
    private lateinit var adapter: ArrayAdapter<String>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, postList)
        post_list_view.adapter = adapter

        testSimpleGeoQuery()
    }

    private fun testSimpleGeoQuery() {
        val singleGeoQuery = geoFirestore.getAtLocation(QUERY_CENTER, QUERY_RADIUS)
        singleGeoQuery.addSingleGeoQueryEventListener(object : SingleGeoQueryDataEventListener {
            override fun onSuccess(documentSnapshots: List<DocumentSnapshot>) {
                documentSnapshots.forEach {
                    val desc = it["DESCRIPTION"] as? String
                    Log.i(TAG, "onSuccess: $desc")
                    adapter.insert(desc, adapter.count)
                    adapter.notifyDataSetChanged()
                }
            }

            override fun onError(exception: Exception) {
                Log.e(TAG, "onError: ", exception)
            }
        })
    }

    private fun testGeoQuery() {
        val geoQuery = geoFirestore.queryAtLocation(QUERY_CENTER, QUERY_RADIUS)
        geoQuery.addGeoQueryDataEventListener(object : GeoQueryDataEventListener {
            override fun onDocumentEntered(documentSnapshot: DocumentSnapshot, location: GeoPoint) {
                val desc = documentSnapshot["DESCRIPTION"] as? String
                Log.i(TAG, "onDocumentEntered: $desc")
                adapter.insert(desc, adapter.count)
                adapter.notifyDataSetChanged()
            }

            override fun onDocumentExited(documentSnapshot: DocumentSnapshot) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onDocumentMoved(documentSnapshot: DocumentSnapshot, location: GeoPoint) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onDocumentChanged(documentSnapshot: DocumentSnapshot, location: GeoPoint) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onGeoQueryReady() {
                Log.i(TAG, "onGeoQueryReady: ")
            }

            override fun onGeoQueryError(exception: Exception) {
                Log.e(TAG, "onGeoQueryError: ", exception)
            }
        })
    }
}